include/nbl/video/alloc/SubAllocatedDescriptorSet.h

```diff
@@ -0,0 +1,120 @@
+	using AddressAllocator = core::GeneralpurposeAddressAllocator<uint32_t>;
+	using ReservedAllocator = core::allocator<uint8_t>;
+	using size_type = typename AddressAllocator::size_type;
+	using value_type = typename AddressAllocator::size_type;
+	static constexpr value_type invalid_value = AddressAllocator::invalid_address;
+
+protected:
+	struct SubAllocDescriptorSetRange {
+		std::shared_ptr<AddressAllocator> addressAllocator;
+		std::shared_ptr<ReservedAllocator> reservedAllocator;
+		size_t reservedSize;
+	};
+	std::vector<SubAllocDescriptorSetRange> m_allocatableRanges = {};
+
+
+public:
+	// constructors
+	template<typename... Args>
+	inline SubAllocatedDescriptorSet(video::IGPUDescriptorSetLayout* layout, const value_type maxAllocatableAlignment, Args&&... args)
+	{
+		for (uint32_t descriptorType = 0; descriptorType < static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT); descriptorType++)
+		{
+			auto descType = static_cast<asset::IDescriptor::E_TYPE>(descriptorType);
+			auto& redirect = layout->getDescriptorRedirect(descType);
+
+			for (uint32_t i = 0; i < redirect.getBindingCount(); i++)
+			{
+				auto binding = redirect.getBinding(i);
+				auto storageIndex = redirect.findBindingStorageIndex(binding);
+
+				auto count = redirect.getCount(storageIndex);
+				auto flags = redirect.getCreateFlags(storageIndex);
+
+				for (uint32_t j = m_allocatableRanges.size(); j < binding.data; j++)
+				{
+					m_allocatableRanges.push_back({});
+				}
+
+				SubAllocDescriptorSetRange range;
+				range.reservedSize = 0;
+				// Only bindings with these flags will be allocatable
+				if (flags.hasFlags(core::bitflag(IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_UPDATE_AFTER_BIND_BIT)
+					| IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_UPDATE_UNUSED_WHILE_PENDING_BIT
+					| IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_PARTIALLY_BOUND_BIT))
+				{
+					range.reservedSize = AddressAllocator::reserved_size(maxAllocatableAlignment, static_cast<size_type>(count), args...);
+					range.reservedAllocator = std::shared_ptr<ReservedAllocator>(new ReservedAllocator());
+					range.addressAllocator = std::shared_ptr<AddressAllocator>(new AddressAllocator(
+						range.reservedAllocator->allocate(range.reservedSize, _NBL_SIMD_ALIGNMENT),
+						static_cast<size_type>(0), 0u, maxAllocatableAlignment, static_cast<size_type>(count), std::forward<Args>(args)...
+					));
+				}
+				m_allocatableRanges.insert(m_allocatableRanges.begin() + binding.data, range);
+			}
+		}
+	}
+
+	~SubAllocatedDescriptorSet()
+	{
+		for (uint32_t i = 0; i < m_allocatableRanges.size(); i++)
+		{
+			auto& range = m_allocatableRanges[i];
+			if (range.reservedSize == 0)
+				continue;
+
+			auto ptr = reinterpret_cast<const uint8_t*>(core::address_allocator_traits<AddressAllocator>::getReservedSpacePtr(*range.addressAllocator));
+			range.reservedAllocator->deallocate(const_cast<uint8_t*>(ptr), range.reservedSize);
```

Comment:
you probably want to force the `addressAllocator` destructor to run first before pulling the reserved memory rug from under its feet